### PR TITLE
[Ethereum] Support decoding contract data / abi

### LINKED ISSUE #1034: [Ethereum] Support parsing common contract calls
common calls like:
ERC20: approve/transfer
ERC721: approve/transferFrom
Kyber: trade
Uniswap v2: swapExact<ETH/TokensFor>

### PR DESCRIPTION
**Pull request by hewigovens**
Fixes #1034.

## Description

Supports Ethereum contract call decoding (types: `uint256`, `address`, `string`, `bool`, `bytes`) return readable json format. C interface also simplified.

A decoded swap call example:

```json
{
    "function": "swapExactTokensForTokens(uint256,uint256,address[],address,uint256)",
    "inputs": [{
        "name": "amountIn",
        "type": "uint256",
        "value": "1000000000000000000"
    }, {
        "name": "amountOutMin",
        "type": "uint256",
        "value": "2494851601099271131"
    }, {
        "name": "path",
        "type": "address[]",
        "value": ["0x6b175474e89094c44da98b954eedeac495271d0f", "0x9f8f72aa9304c8b593d555f12ef6589cc3a579a2", "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2", "0xe41d2489571d322189246dafa5ebde1f4699f498"]
    }, {
        "name": "to",
        "type": "address",
        "value": "0x7d8bf18c7ce84b3e175b339c4ca93aed1dd166f1"
    }, {
        "name": "deadline",
        "type": "uint256",
        "value": "1594806384"
    }]
}
```

TODO:
- [x] Swift test
- [x] Kotlin test
- [x] TW interface test

## Testing instructions

Unit tests

## Types of changes

* Breaking change

### CODE REVIEW

**Comment by hewigovens** [comment]
@vikmeup https://github.com/trustwallet/wallet-core/tree/xt/contract-parsing/tests/Ethereum/Data what else should we support?

File: src/Ethereum/ContractCall.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+#include "uint256.h"
+#include <iostream>
+
+using namespace TW;
+using namespace TW::Ethereum;
+
+static void fill(ABI::Function& func, const std::string& type) {
+    if (type == "address") {
+        auto param = std::make_shared<ABI::ParamAddress>();
+        func.addParam(param, false);
```
Comment (catenocrypt) [inline src/Ethereum/ContractCall.cpp]: The 'func.addParam' line is duplicated 5 times.

File: src/Ethereum/ABI/ParamFactory.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    } else if (type == "uint256") {
+        param = make_shared<ParamUInt256>();
+    } else if (type == "bool") {
+        param = make_shared<ParamBool>();
+    } else if (type == "bytes") {
+        param =  make_shared<ParamByteArray>();
+    } else if (type == "string") {
+        param = make_shared<ParamString>();
+    }
+    return param;
```
Comment (catenocrypt) [inline src/Ethereum/ABI/ParamFactory.cpp:27]: What if type is none of the above?

File: src/Ethereum/ABI/ParamFactory.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        auto value = dynamic_pointer_cast<ParamBool>(param);
+        result = value->getVal() ? "true" : "false";
+    } else if (type == "bytes") {
+        auto value = dynamic_pointer_cast<ParamByteArray>(param);
+        result = hexEncoded(value->getVal());
+    }  else if (type == "string") {
+        auto value = dynamic_pointer_cast<ParamString>(param);
+        result = value->getVal();
+    }
+    return result;
```
Comment (catenocrypt) [inline src/Ethereum/ABI/ParamFactory.cpp:48]: What if type is none of the above?

File: src/Ethereum/ContractCall.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+static void fillArray(Function& func, const string& type) {
+    auto param = make_shared<ParamArray>();
+    auto baseType = string(type.begin(), type.end() - 2);
+    auto value = ParamFactory::make(baseType);
+    param->addParam(value);
+    func.addParam(param, false);
+}
+
+static void fill(Function& func, const string& type) {
+    if (boost::algorithm::ends_with(type, "[]")) {
```
Comment (catenocrypt) [inline src/Ethereum/ContractCall.cpp:29]: Std::string has ends_with, refrain from boost when not needed.
https://en.cppreference.com/w/cpp/string/basic_string/ends_with

File: src/Ethereum/ContractCall.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        return fillArray(func, type);
+    } else {
```
Comment (catenocrypt) [inline src/Ethereum/ContractCall.cpp:31]: Superfluous else after return.

File: src/Ethereum/ContractCall.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        auto param = ParamFactory::make(type);
+        func.addParam(param, false);
+    }
+}
+
+static vector<string> getArrayValue(Function& func, const string& type, int idx) {
+    auto baseType = string(type.begin(), type.end() - 2);
+    shared_ptr<ParamBase> param;
+    func.getInParam(idx, param);
+    auto array = dynamic_pointer_cast<ParamArray>(param);
```
Comment (catenocrypt) [inline src/Ethereum/ContractCall.cpp:41]: If param is up-cast to ParamArray, why isn't it created ParamArray from start?

File: src/Ethereum/ContractCall.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+static string getValue(Function& func, const string& type, int idx) {
+    shared_ptr<ParamBase> param;
+    func.getInParam(idx, param);
+    return ParamFactory::getValue(param, type);
+}
+
+static json buildInputs(Function& func, const json& registry) {
+    auto inputs = json::array();
+    for (int i = 0; i < registry["inputs"].size(); i++) {
```
Comment (catenocrypt) [inline src/Ethereum/ContractCall.cpp:58]: Error handling for missing "inputs"?

File: src/Ethereum/ContractCall.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        auto info = registry["inputs"][i];
+        auto type = info["type"];
+        auto input = json{
+            {"name", info["name"]},
+            {"type", type}
+        };
+        if (boost::algorithm::ends_with(type.get<string>(), "[]")) {
```
Comment (catenocrypt) [inline src/Ethereum/ContractCall.cpp:65]: Use string::ends_with

File: src/Ethereum/ContractCall.cpp [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+#include <boost/algorithm/string/predicate.hpp>
+#include <iostream>
+
+using namespace TW;
+using namespace TW::Ethereum::ABI;
+using namespace std;
+
+using json = nlohmann::json;
+
+static void fillArray(Function& func, const string& type) {
```
Comment (catenocrypt) [inline src/Ethereum/ContractCall.cpp:20]: Private function in global namespace; better to use namespace TW::Ethereum::ABI for whole file.

File: src/Ethereum/ContractCall.h [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+// file LICENSE at the root of the source code distribution tree.
+
+#pragma once
+
+#include "Data.h"
+#include <nlohmann/json.hpp>
+#include <optional>
+#include <string>
+
+namespace TW::Ethereum {
```
Comment (catenocrypt) [inline src/Ethereum/ContractCall.h:14]:
```suggestion
namespace TW::Ethereum::ABI {
```

File: tests/Ethereum/ContractCallTests.cpp [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+#include "Ethereum/ContractCall.h"
+#include "HexCoding.h"
+
+#include <fstream>
+#include <gtest/gtest.h>
+
+using namespace TW;
+using namespace TW::Ethereum;
+
+extern std::string TESTS_ROOT;
```
Comment (catenocrypt) [inline tests/Ethereum/ContractCallTests.cpp:16]: Why using extern here?  Where is it set?

**Comment by catenocrypt** [comment]
Diff coverage is 85.80% -- can it be improved?

File: src/Ethereum/ABI/contracts/erc20.json [hunk 12]
```diff
@@ -0,0 +1,50 @@
+{
```
[the remark this answers is not in the document]
Comment (hewigovens) [reply src/Ethereum/ABI/contracts/erc20.json]: moved test folder, it seems that you reviewed old code

File: tests/Ethereum/ContractCallTests.cpp (hunk 11 above)
Comment (hewigovens) [reply tests/Ethereum/ContractCallTests.cpp:16]: in gtest main.cpp

**Comment by hewigovens** [comment]
@catenocrypt will add more C interface tests tmr

File: src/Ethereum/ContractCall.cpp (hunk 4 above)
Comment (hewigovens) [reply src/Ethereum/ContractCall.cpp:29]: that's in c++20, I'm not sure if iOS/Android toolchain has it

File: src/Ethereum/ABI/ParamFactory.cpp (hunk 2 above)
Comment (hewigovens) [reply src/Ethereum/ABI/ParamFactory.cpp:27]: Will only add most common used types first, what else should we add in first batch?

File: src/Ethereum/ContractCall.cpp (hunk 6 above)
Comment (hewigovens) [reply src/Ethereum/ContractCall.cpp:41]: I got this error, do you have better way to get concrete type?

```shell
cannot bind to a value of unrelated type 'shared_ptr<TW::Ethereum::ABI::ParamArray>'
func.getInParam(idx, param);
```

File: src/Ethereum/ContractCall.cpp (hunk 7 above)
Comment (hewigovens) [reply src/Ethereum/ContractCall.cpp:58]: we won't decode arbitrary abi json tho, need to pre calculate method id + correct abi json

File: src/Ethereum/ContractCall.cpp (hunk 5 above)
Comment (hewigovens) [reply src/Ethereum/ContractCall.cpp:30]: done

File: src/Ethereum/ContractCall.cpp (hunk 9 above)
Comment (hewigovens) [reply src/Ethereum/ContractCall.cpp:19]: done

File: src/Ethereum/ContractCall.h (hunk 10 above)
Comment (hewigovens) [reply src/Ethereum/ContractCall.h]: done

File: src/Ethereum/ContractCall.cpp (hunk 8 above)
Comment (hewigovens) [reply src/Ethereum/ContractCall.cpp:64]: Just checked Xcode, it has some `c++2a` features support but no explicit settings like `c++17`, we can enable c++2a once it's official supported

File: android/build.gradle [hunk 13]
```diff
@@ -7,7 +7,7 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.6.3'
+        classpath 'com.android.tools.build:gradle:4.0.0'
```
Comment (madcake) [inline android/build.gradle:10]: Latest is 4.0.1

File: src/Ethereum/ABI/ParamFactory.cpp (hunk 2 above)
Comment (hewigovens) [reply src/Ethereum/ABI/ParamFactory.cpp:56]: most types are handled now

File: android/build.gradle (hunk 13 above)
Comment (hewigovens) [reply android/build.gradle]: done

**Pull request merged**